### PR TITLE
Governor and token implementation

### PR DESCRIPTION
**Pull request by JasperTimm**
This PR includes the token and token factory branch from this PR: #8 

It therefore covers issues: #4, #5 and #6

### CODE REVIEW

**Comment by JasperTimm** [comment]
> Hey!
> 
> _function createProposal_
> lines:
> _governor.proposalCount++; governor.proposalList[governor.proposalCount] = newProposal;_
> 
> governor.proposalCount and governor.proposalList are designed to be ever-growing? Or they stores only **active** proposals?
> 
> a. if ever-growing, do we want the governor to be the object that stores all the history active/inactive proposals...
> 
> б. if stores only active, found some vulnerability in these two lines - proposals can overlap.
> Imagine the case:
> 
> 1. I create new proposal. "governor.proposalCount" turns 5. My new proposal is №5 in "governor.proposalCount".
> 2. Later on, myProposal №3 is over (somehow). "governor.proposalCount" should turn 4.
> 3. I create one more new proposal. "governor.proposalCount" turns 5. This new proposal takes place №5 in "governor.proposalCount", that has already been taken in Step1.
> 
> What do you think about the trade-off?
> 
> * governor.proposalCount to be ever-growing,
> * governor.proposalList to be storing only active proposals

Hey @daniilogurtsov - sorry I didn't reply to this earlier. Thanks for taking a look at the commitments, it's good to know someone is checking code for vulnerabilities. For this case, I had always considered that it would be an ever expanding list of proposals where we keep past ones around, probably this is a good idea for the sake of the UI as well so we can see the history of past failed or successful proposals.

File: contracts/OctobayGovTokenFactory.sol [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+pragma solidity ^0.6.0;
+pragma experimental ABIEncoderV2;
+
+import './OctobayStorage.sol';
+import './OctobayGovToken.sol';
+
+contract OctobayGovTokenFactory is OctobayStorage {
+
+    event NewTokenEvent(string name, string symbol, address tokenAddr);
+    event UpdatedProjectId(string oldProjectId, string newProjectId);
```
Comment (mktcode) [inline contracts/OctobayGovTokenFactory.sol]: should also include the token address, like in the line above

File: contracts/OctobayGovernor.sol [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+    struct Vote {
+        bool hasVoted; // to track whether a token holder has already voted
+        int16 vote; // can be negative, represents token share (0 - 10000)
+    }
+
+    enum ProposalState {
+        Pending,
+        Active,
+        Defeated,
```
Comment (mktcode) [inline contracts/OctobayGovernor.sol:40]: uuhhh you mean like... having too many "downvotes" could trigger a negative quorum, so to say? yeah let's do that.

File: contracts/OctobayGovToken.sol [hunk 3]
```diff
@@ -0,0 +1,45 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.6.0;
+pragma experimental ABIEncoderV2;
+
+import '@openzeppelin/contracts/token/ERC20/ERC20.sol';
+import './OctobayStorage.sol';
+
+contract OctobayGovToken is OctobayStorage, ERC20 {
+
```
Comment (mktcode) [inline contracts/OctobayGovToken.sol:9]: I think here we'll then need properties like `minProposalSuccessQuorum`, `minProposalRejectQuorum` and `minShareForProposing`

File: contracts/OctobayGovTokenFactory.sol [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    mapping (string => OctobayGovToken) public tokensByProjectId;
+
+    /// @param _name Name of the new token
+    /// @param _symbol Token Symbol for the new token
+    /// @param _projectId Path of the org or repo which maps to the new token
+    /// @return The address of the new token contract
+    function createToken(string memory _name, string memory _symbol, string memory _projectId) external onlyOctobay returns (OctobayGovToken) {
```
Comment (mktcode) [inline contracts/OctobayGovTokenFactory.sol:18]: needs implementation of those parameters mentioned above then.

Comment (mktcode) [reply contracts/OctobayGovTokenFactory.sol:18]: and they need to be updateable by owner (and later also by the owner-nft holder)

**Comment by mktcode** [comment]
@daniilogurtsov

> What do you think about the trade-off?
> 
>     * governor.proposalCount to be ever-growing,
> 
>     * governor.proposalList to be storing only active proposals

An ever-growing list is fine I guess, at least for now. Removing data from the blockchain safes gas-cost though, so it's always a thing to consider.

@JasperTimm For the frontend it could actually be removed from the contract as long as it's still in our subgraph. Did I mention today how much I love the graph? :smile: 

Also from my side... sorry for the late response. I promise betterment! :100: % focused on the hackathon now.

Oh and btw @daniilogurtsov

> _function createProposal_
> lines:
> _governor.proposalCount++; governor.proposalList[governor.proposalCount] = newProposal;_

There's a nice code commenting feature. You see it in action above. When you go to the "files changed" tab at the top, you'll find it.

ezgif-3-a3b00a176012

**Comment by mktcode** [comment]
@JasperTimm I think we will need this functionality right from the start: https://github.com/Giveth/minime/blob/master/contracts/MiniMeToken.sol#L283

When votes are cast we should look at the token share at the time the proposal was made, don't you think?

File: contracts/OctobayGovTokenFactory.sol (hunk 1 above)
Comment (JasperTimm) [reply contracts/OctobayGovTokenFactory.sol]: Fixed

File: contracts/OctobayGovToken.sol (hunk 3 above)
Comment (JasperTimm) [reply contracts/OctobayGovToken.sol:9]: Given what we discussed now, I guess these aren't required now?

**Comment by JasperTimm** [comment]
I'm going to work on the balanceAtBlock functionality now, as I think that's the only outstanding change here for now.

File: contracts/OctobayGovToken.sol (hunk 3 above)
Comment (mktcode) [reply contracts/OctobayGovToken.sol:9]: true

Comment (mktcode) [reply contracts/OctobayGovToken.sol:9]: well... `minShareForProposing` still is.

**Comment by JasperTimm** [comment]
@mktcode - is there anything outstanding on this?

**Comment by mktcode** [comment]
If so, we'll handle it in a separate issue.

**Review by mktcode — APPROVED** [review]

**Pull request merged**